[MLP-195] Api for creating a user

darkainsoul opened the pull request:
## Background
API method that uses Agatha as a Request/Response Service Layer for creating one User, a model was used to keep the flow of the API until de Datalayer
## Changes done
- Add CRUD for users
- Add Test project
## Pending to be done
Refactor the UsersService to connect with data layer
## Notes
The nuget packages sometimes not be restored immediately for this case just rebuild the solution

martinkeimel commented on P202-Training.Domain/Models/Role.cs:
```diff
@@ -0,0 +1,13 @@
+﻿using System;
```
Please review unused usings

martinkeimel commented on .vs/config/applicationhost.config:
```diff
@@ -0,0 +1,1047 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Please do not push this file, as I've recently added to gitignore

martinkeimel commented on .vs/ProjectSettings.json:
```diff
@@ -0,0 +1,3 @@
+{
```
What's this file for?

martinkeimel commented on .vs/VSWorkspaceState.json:
```diff
@@ -0,0 +1,6 @@
+{
```
What's this file for?

martinkeimel commented on .vs/config/applicationhost.config.orig:
```diff
@@ -0,0 +1,1047 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Please remove this file. Every file ending with .orig is a temporary file used for solving merge conflicts

martinkeimel commented on P202-Training.Domain/UsersService.cs:
```diff
@@ -0,0 +1,55 @@
+        public void CreateUser(Users users)
+        {
+            if (users == null) return;
+            Mapper.Initialize(cfg => cfg.CreateMap<Users, User>());
```
It's ok if you want to use Automapper but in that case, mapping should be configured only once. Perhaps the global.asax it's a good place to do that. And the mapper should be injected as a dependency to each of the classes that make use of it

martinkeimel commented on P202-Training.Domain/Models/Users.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace P202.Training.Domain.Models
+{
+    public class Users
```
Should this class be called `User`? (in singular)

martinkeimel commented on P202-Training.Domain/UsersService.cs:
```diff
@@ -0,0 +1,55 @@
+            if (users == null) return;
+            Mapper.Initialize(cfg => cfg.CreateMap<Users, User>());
+            var mapUser = Mapper.Map<User>(users);
+            var userRepository = new UserRepository(new NHibernateSessionManager());
```
Same here, you sholdn't be creating a new `UserRepository`, you should inject the `UserRepository` and let the DI engine handle object creation

martinkeimel commented on P202-Training.Domain/UsersService.cs:
```diff
@@ -0,0 +1,55 @@
+
+        public Users ReadUser(int userId)
+        {
+            return new Users()
```
If you're gonna implement this method in a separated branch it's OK, but please leave a TODO comment

martinkeimel commented on P202-Training.Domain/UsersService.cs:
```diff
@@ -0,0 +1,55 @@
+        }
+
+        public Users UpdateUser(Users users)
+        {
```
If you're gonna implement this method in a separated branch it's OK, but please leave a TODO comment

martinkeimel commented on P202-Training.Domain/UsersService.cs:
```diff
@@ -0,0 +1,55 @@
+
+        public IList<Users> ListUsers()
+        {
+            var userRepository = new UserRepository(new NHibernateSessionManager());
```
Same here: inject Mapper and UserRepository

martinkeimel commented on P202.Training.Test/P202.Training.Test.csproj:
```diff
@@ -0,0 +1,89 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Thanks A LOT for adding the test project 😉

martinkeimel commented on P202.Training.Test/UsersServiceTest.cs:
```diff
@@ -0,0 +1,58 @@
+    public class UsersServiceTest
+    {
+        [Fact]
+        public void UsersService_CreateData_WithoutDatabase()
```
If you inject the repository as a dependency, it shouldn't matter if you're using a database or not... you'd be mocking the repository methods and test just the domain layer logic

martinkeimel replied:
Same goes for the rest of the tests. And also, all the tests should have at least one assertion

martinkeimel commented on P202.Training.WCF/Handlers/DeleteUserHandler.cs:
```diff
@@ -0,0 +1,28 @@
+﻿using System;
```
Please remove unused usings

martinkeimel commented on P202.Training.WCF/Handlers/ListUsersHandler.cs:
```diff
@@ -0,0 +1,28 @@
+﻿using System;
```
Please remove unused usings

martinkeimel commented on P202.Training.WCF/Handlers/ReadUserHandler.cs:
```diff
@@ -0,0 +1,28 @@
+﻿using System;
```
Please remove unused usings

martinkeimel commented on P202.Training.WCF/Handlers/UpdateUserHandler.cs:
```diff
@@ -0,0 +1,28 @@
+﻿using System;
```
Please remove unused usings

martinkeimel commented on P202.Training.WCF/P202.Training.WCF.csproj:
```diff
@@ -30,6 +31,8 @@
     <DefineConstants>DEBUG;TRACE</DefineConstants>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
+    <CodeAnalysisRuleSet>..\..\..\..\Downloads\CSharpStyleguide.ruleset</CodeAnalysisRuleSet>
```
I think you're making reference to a path in your local disk that is not present in the repository. It's ok if you want to add this ruleset file in the root of the project and make a reference to it

martinkeimel commented on P202.Training.WCF/P202.Training.WCF.csproj.orig:
```diff
@@ -0,0 +1,185 @@
+﻿<Project ToolsVersion="15.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Please remove this orig file

martinkeimel commented on P202.Training.WCF/RequestsAndResponses/DeleteUserRequest.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using System;
```
Please remove unused usings

martinkeimel commented on P202.Training.WCF/RequestsAndResponses/DeleteUserResponse.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using System;
```
Please remove unused usings

martinkeimel commented on P202.Training.WCF/RequestsAndResponses/CreateUserRequest.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public class CreateUserRequest : Request
+    {
+        public Users Value { get; set; }
```
Would you mind renaming the `Value` property to sth more meaningful? Perhaps `User` or `NewUser`

martinkeimel commented on P202.Training.WCF/RequestsAndResponses/CreateUserResponse.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public class CreateUserResponse : Response
+    {
+        public Users Value { get; set; }
```
Same here: rename value property

martinkeimel commented on P202.Training.WCF/RequestsAndResponses/ListUsersRequest.cs:
```diff
@@ -0,0 +1,13 @@
+﻿using System;
```
Please remove unused usings

martinkeimel commented on P202.Training.WCF/RequestsAndResponses/ListUsersResponse.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using System;
```
Please remove unused usings

martinkeimel commented on P202.Training.WCF/RequestsAndResponses/ReadUserRequest.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using System;
```
Please remove unused usings

martinkeimel commented on P202.Training.WCF/RequestsAndResponses/ReadUserResponse.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using System;
```
Please remove unused usings

martinkeimel commented on P202.Training.WCF/RequestsAndResponses/UpdateUserRequest.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using System;
```
Please remove unused usings

martinkeimel commented on P202.Training.WCF/RequestsAndResponses/UpdateUserResponse.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using System;
```
Please remove unused usings

martinkeimel commented on P202.Training.WCF/RequestsAndResponses/DeleteUserRequest.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public class DeleteUserRequest : Request
+    {
+        public int Value { get; set; }
```
Same here: rename value property

martinkeimel commented on P202.Training.WCF/RequestsAndResponses/DeleteUserResponse.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public class DeleteUserResponse : Response
+    {
+        public Users Value { get; set; }
```
Same here: rename value property

martinkeimel commented on P202.Training.WCF/RequestsAndResponses/ListUsersResponse.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public class ListUsersResponse : Response
+    {
+        public IList<Users> Value { get; set; }
```
Same here: rename value property

martinkeimel commented on P202.Training.WCF/RequestsAndResponses/ReadUserRequest.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public class ReadUserRequest : Request
+    {
+        public int Value { get; set; }
```
Same here: rename value property

martinkeimel commented on P202.Training.WCF/RequestsAndResponses/ReadUserResponse.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public class ReadUserResponse : Response
+    {
+        public Users Value { get; set; }
```
Same here: rename value property

martinkeimel commented on P202.Training.WCF/RequestsAndResponses/UpdateUserRequest.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public class UpdateUserRequest : Request
+    {
+        public Users Value { get; set; }
```
Same here: rename value property

martinkeimel commented on P202.Training.WCF/RequestsAndResponses/UpdateUserResponse.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public class UpdateUserResponse : Response
+    {
+        public Users Value { get; set; }
```
Same here: rename value property

martinkeimel commented on P202.Training.WCF/packages.config.orig:
```diff
@@ -0,0 +1,16 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Please remove orig file

martinkeimel reviewed (commented):
I left a lot of comments but don't worry, most of them are related to code styling. The most important thing you should pay attention to is the comments related to DI.

Also, I see you made a merge before rebasing. That may bring you troubles when you want to squash this branch so let me know if I can help you with that rebase.

Overall looks awesome! :1st_place_medal:

darkainsoul committed:
Integration with Datalayer for
- MLP-195 [P202] Create API method for creating one User
- MLP-202 [P202] Create API method for getting all the Users

Refactor of code styles

darkainsoul committed:
Fix code styles

martinkeimel commented on .gitignore:
```diff
@@ -29,4 +29,4 @@ _ReSharper*/
 packages/
 .vs/config/applicationhost.config
-webapp/node_modules/**/*
+.vs/
```
Please do not remove the ignore for the node_modules folder. You should leave both ignore lines

martinkeimel commented on P202-Training.Domain/IUsersService.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    public interface IUsersService
+    {
+        void CreateUser(User users, IUserRepository userRepository);
```
You shouldn't be sending the repository as a parameter in each method. That should be injected in the service constructor. And it definitely shouldn't be part of the interface

martinkeimel commented on P202-Training.Domain/UsersService.cs:
```diff
@@ -0,0 +1,49 @@
+        public void CreateUser(User users, IUserRepository userRepository)
+        {
+            if (users == null) return;
+            var mapUser = Mapper.Map<Data.Entities.User>(users);
```
You're still using the Mapper static class. Is there any IMapper interface that you can inject to this class? If not perhaps you can create one that acts as a wrapper to the Mapper static class

martinkeimel commented on P202-Training.Domain/UsersService.cs:
```diff
@@ -0,0 +1,49 @@
+
+        public void DeleteUser(int userId, IUserRepository userRepository)
+        {
+            // TODO
```
It's ok if you wan't to leave a TODO comment but please add some more information (ie: // TODO: Implement DeleteUser method)

martinkeimel commented on P202.Training.Test/UsersServiceTest.cs:
```diff
@@ -0,0 +1,43 @@
+        public void UsersService_CreateData()
+        {
+            IUsersService usersService = new UsersService();
+            // TODO
```
Same comment for the TODOs

martinkeimel commented on P202.Training.WCF/Handlers/CreateUserHandler.cs:
```diff
@@ -0,0 +1,27 @@
+    public class CreateUserHandler : RequestHandler<CreateUserRequest, CreateUserResponse>
+    {
+        private readonly IUsersService _usersService;
+        private readonly IUserRepository _userRepository;
```
API layer shouldn't have a reference to the repository. Repositories should be accessed only from the Domain layer

martinkeimel reviewed (commented):
DI still needs some more work. Also there are some orig files still present in the branch. Let me know if you need any help

darkainsoul committed:
Change static Automapper for IMapper
Remove IRepository from IUsersService

darkainsoul committed:
Remove .orig files

martinkeimel reviewed (approved):
Looks great! 💯
